javalib/src/main/scala/java/util/Arrays.scala

```diff
@@ -12,398 +12,411 @@
-    binarySearchImpl[Short](a, 0, a.length, key, _ < _)
+    binarySearchImpl(a, 0, a.length, key)(ShortArrayOps)
 
   @noinline def binarySearch(a: Array[Short], startIndex: Int, endIndex: Int, key: Short): Int = {
     checkRangeIndices(a, startIndex, endIndex)
-    binarySearchImpl[Short](a, startIndex, endIndex, key, _ < _)
+    binarySearchImpl(a, startIndex, endIndex, key)(ShortArrayOps)
   }
 
   @noinline def binarySearch(a: Array[Char], key: Char): Int =
-    binarySearchImpl[Char](a, 0, a.length, key, _ < _)
+    binarySearchImpl(a, 0, a.length, key)(CharArrayOps)
 
   @noinline def binarySearch(a: Array[Char], startIndex: Int, endIndex: Int, key: Char): Int = {
     checkRangeIndices(a, startIndex, endIndex)
-    binarySearchImpl[Char](a, startIndex, endIndex, key, _ < _)
+    binarySearchImpl(a, startIndex, endIndex, key)(CharArrayOps)
   }
 
   @noinline def binarySearch(a: Array[Byte], key: Byte): Int =
-    binarySearchImpl[Byte](a, 0, a.length, key, _ < _)
+    binarySearchImpl(a, 0, a.length, key)(ByteArrayOps)
 
   @noinline def binarySearch(a: Array[Byte], startIndex: Int, endIndex: Int, key: Byte): Int = {
     checkRangeIndices(a, startIndex, endIndex)
-    binarySearchImpl[Byte](a, startIndex, endIndex, key, _ < _)
+    binarySearchImpl(a, startIndex, endIndex, key)(ByteArrayOps)
   }
 
   @noinline def binarySearch(a: Array[Double], key: Double): Int =
-    binarySearchImpl[Double](a, 0, a.length, key, _ < _)
+    binarySearchImpl(a, 0, a.length, key)(DoubleArrayOps)
 
   @noinline def binarySearch(a: Array[Double], startIndex: Int, endIndex: Int, key: Double): Int = {
     checkRangeIndices(a, startIndex, endIndex)
-    binarySearchImpl[Double](a, startIndex, endIndex, key, _ < _)
+    binarySearchImpl(a, startIndex, endIndex, key)(DoubleArrayOps)
   }
 
   @noinline def binarySearch(a: Array[Float], key: Float): Int =
-    binarySearchImpl[Float](a, 0, a.length, key, _ < _)
+    binarySearchImpl(a, 0, a.length, key)(FloatArrayOps)
 
   @noinline def binarySearch(a: Array[Float], startIndex: Int, endIndex: Int, key: Float): Int = {
     checkRangeIndices(a, startIndex, endIndex)
-    binarySearchImpl[Float](a, startIndex, endIndex, key, _ < _)
+    binarySearchImpl(a, startIndex, endIndex, key)(FloatArrayOps)
   }
 
   @noinline def binarySearch(a: Array[AnyRef], key: AnyRef): Int =
-    binarySearchImplRef(a, 0, a.length, key)
+    binarySearchImpl(a, 0, a.length, key)(NaturalComparator)
 
   @noinline def binarySearch(a: Array[AnyRef], startIndex: Int, endIndex: Int, key: AnyRef): Int = {
     checkRangeIndices(a, startIndex, endIndex)
-    binarySearchImplRef(a, startIndex, endIndex, key)
+    binarySearchImpl(a, startIndex, endIndex, key)(NaturalComparator)
   }
 
   @noinline def binarySearch[T <: AnyRef](a: Array[T], key: T, c: Comparator[_ >: T]): Int =
-    binarySearchImpl[T](a, 0, a.length, key, (a, b) => c.compare(a, b) < 0)
+    binarySearchImpl[T](a, 0, a.length, key)(ifNullUseNaturalComparator(c))
```

Comment:
Seems like this was also broken (throws NPE if `c == null`). Add a test case?

Reply:
Ah, indeed. I extracted that fix in a separate commit.